### PR TITLE
Make use alongside standard implementations safe

### PR DESCRIPTION
**Pull request by nmcclatchey**
- Moves the implementation to its own namespace, `mingw_stdthread`, and removes the `#error` that occurs when including the implementation alongside GCC's implementation. This makes simultaneous inclusion of both implementations safe.
- Removes the namespace inclusion trick from `mingw.shared_mutex.h`. No ambiguity occurred when objects specified a namespace with `std::`, but `using namespace std;` fully flattens the namespace tree, which could have introduced ambiguity.
- Added a check for the presence of MinGW before including files into `std`. This should allow the library to be used as-is with Visual Studio, etc., but might cause `custom_namespace::std` to be empty if the header is included inside `custom_namespace`. To achieve compatibility with legacy code with minimal changes, one could rename the `mingw.*.h` headers and create new headers like the following:
`#include "mingw.*.h"
namespace std {
using namespace mingw_stdthread;
}`.
These compatibility headers would clash with namespace `std` (if `using namespace std;` is declared, and if not included in their own namespace), but can safely be included inside a custom namespace.
- Fixes a test in `tests.cpp`. The original test would fail unless an object was moved exactly once. The standard library moves the object multiple times, which would cause the test to fail. The new test allows repeated movement.

### CODE REVIEW

File: mingw.condition_variable.h [hunk 1]
```diff
@@ -20,34 +20,24 @@
...
+//  Use the standard classes for std::, if available.
+#include <condition_variable>
+
 #include <atomic>
 #include <assert.h>
 #include "mingw.mutex.h"
 #include <chrono>
 #include <system_error>
 #include <windows.h>
-#ifdef _GLIBCXX_HAS_GTHREADS
```
Comment (alxvasilev) [inline mingw.condition_variable.h]: I would like to preserve this as a warning

File: mingw.thread.h [hunk 2]
```diff
@@ -30,22 +33,9 @@
 #include <process.h>
 #include <ostream>
 
-#ifdef _GLIBCXX_HAS_GTHREADS
-#error This version of MinGW seems to include a win32 port of pthreads, and probably    \
```
Comment (alxvasilev) [inline mingw.thread.h]: I would like to preserve this as a warning

File: mingw.thread.h [hunk 3]
```diff
@@ -20,6 +20,9 @@
 #ifndef WIN32STDTHREAD_H
 #define WIN32STDTHREAD_H
 
+//  Use the standard classes for std::, if available.
```
[the remark this answers is not in the document]
Comment (nmcclatchey) [reply mingw.thread.h]: I operate under the assumption that a user who includes mingw.thread.h desires that the namespace std contain the standard threading classes.
- If those are not already provided by the compiler, the mingw-std-thread library adds them.
- If those are already provided by the compiler, then we cannot both place the mingw-std-thread library's classes into std and have them yield to those of <thread>. The namespace inclusion trick that would have allowed this is foiled by the common command `using namespace std;`.
  - If we `#include <thread>`, then a user need only `#include "mingw.thread.h`, regardless of which compiler is used.
 - If we do not `#include <thread>`, then a user must independently specify the use of <thread> if their code is used with a different compiler.

In short, using `#include <thread>` in the mingw-std-thread library doesn't hurt, and simplifies the experience of the end user.

File: tests/tests.cpp [hunk 4]
```diff
@@ -73,8 +73,10 @@ int main()
         try
         {
             LOG("Worker thread started, sleeping for a while...");
-            assert(a.mStr == "move test moved" && !strcmp(b, "test message")
-               && (c == -20));
+//  Thread might move the string more than once.
```
[the remark this answers is not in the document]
Comment (nmcclatchey) [reply tests/tests.cpp]: I have been testing on both the win32 and posix versions of MinGW w64, as part of ensuring that the library doesn't cause conflicts. The multiple move happens when using winpthreads (due to the "use existing implementation" part of my contribution working as intended). I suspect it might also happen if the test is run on Visual Studio, though I have not tested that case.

Multiple moves are made apparent by your code with or without the assertion. Expect "move test moved moved moved" when using libwinpthreads.

File: mingw.condition_variable.h (hunk 1 above)
Comment (nmcclatchey) [reply mingw.condition_variable.h]: See #33.

File: mingw.thread.h (hunk 2 above)
Comment (nmcclatchey) [reply mingw.thread.h]: See #33.

File: mingw.thread.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (alxvasilev) [reply mingw.thread.h]: Ok lets leave it like that, but generally I would avoid doing things behind the scenes that the user may not intend to do, for whatever reason, even if it would save a line or two of code.

File: tests/tests.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (alxvasilev) [reply tests/tests.cpp]: Okay, so this is related to the silent inclusion of <thread> if it is available. This assert was intended to test the efficiency of our implementation, not the standard lib. There was a PR from another user that raised the topic about the efficiency of passing the arguments, which was done via std::bind. With his help I managed to optimize that without using C++14 features. It seems that implementation is more efficient in that it does only a single move. Not sure if that would work on all compilers, so that strict assert is not a good idea
 But it raises another point - we have to be sure which implementation is actually being tested.

**Comment by alxvasilev** [comment]
I have merged this PR, but created another one that refines the handling of compiler and std::thread existence in #35. Please have a look and leave comments in that PR.
